### PR TITLE
Reimplement TCP Server with libuv

### PR DESCRIPTION
**Pull request by thestr4ng3r**
Please review @radare

See libr/core/rtr.c for the relevant changes.
`.:1337` works really well already with libuv including running in a task, breaking and so on.

I've added the libuv source to shlr and build it in meson. acr+make still has to be done and meson also needs some additions for platforms other than Linux and Windows.
I am aware it is quite a large new dependency, but it makes the server code WAY better and safer.

### CODE REVIEW

**Comment by ret2libc** [comment]
Please rebase on top of master

**Comment by radare** [comment]
libuv is an actively maintained library, i dont see the reason to put it in shlr/, also i dont want this to be a heavy dependency on r2. so i would prefer to have it as an option at compile time instead of replacing the current code

**Comment by radare** [comment]
and squash

**Comment by radare** [comment]
but i dont want to add more things into shlr/ unless they are unmaintained and we need to fork them

File: libr/core/rtr.c [hunk 1]
```diff
@@ -1847,68 +1848,197 @@ R_API char *r_core_rtr_cmds_query (RCore *core, const char *host, const char *po
...
+	size_t len;
+	uv_tcp_t *client;
+} rtr_cmds_client_context;
+
+static void rtr_cmds_client_close(uv_tcp_t *client, bool remove) {
+	uv_loop_t *loop = client->loop;
+	rtr_cmds_context *context = loop->data;
+	if (remove) {
+		size_t i;
+		for (i=0; i<r_pvector_len (&context->clients); i++) {
```
Comment (radare) [inline libr/core/rtr.c]: spaces around =

File: libr/core/rtr.c [hunk 2]
```diff
@@ -1847,68 +1848,197 @@ R_API char *r_core_rtr_cmds_query (RCore *core, const char *host, const char *po
...
+	}
+
+	if (!client_context->res || (!r_config_get_i (client_context->core->config, "scr.prompt") &&
+				 !strcmp ((char *)buf, "q!")) ||
+				 !strcmp ((char *)buf, ".--")) {
+		rtr_cmds_client_close ((uv_tcp_t *) client, true);
+		return;
+	}
+
+	uv_write_t *req = R_NEW (uv_write_t);
```
Comment (radare) [inline libr/core/rtr.c]: R_NEW0 ?

**Comment by radare** [comment]
for me its ok to merge this if we keep the old code and make libuv an optional dependency and outside shlr/

**Comment by thestr4ng3r** [comment]
Yes I totally agree. I am not a fan of the copied source code in shlr too. With meson it's trivial to find libuv from system or somewhere else, I'll have to look into how to do it with acr.

However, I think we should kill the old server code completely, because it's crap compared to the libuv solution. First, the breaking gets broken when using more than one server at a time and there are also some other issues like not handling tcp packet fragmentation correctly.
What would you think about making the servers an optional feature depending on libuv altogether?

**Comment by radare** [comment]
With acr just use PKGCFG directive. Its 1 line, and then 2 lines in the makefile

**Comment by thestr4ng3r** [comment]
Sure, that won't be a problem, but please answer this:

> However, I think we should kill the old server code completely, because it's crap compared to the libuv solution: the breaking gets broken when using more than one server at a time and there are also some other issues like not handling tcp packet fragmentation correctly.
What would you think about making the servers an optional feature depending on libuv altogether?

**Comment by radare** [comment]
Yes i know, doing async/coroutined networking is not easy, and libuv does that pretty well, in exchange of lot of boilerplate and layers of abstractions, i want to keep the current implementation because it's less than 400 lines and works without any dependency. which is useful in many situations where you dont care about performance or threads at all like embedded devices with low memory constraints.

File: libr/core/rtr.c (hunk 2 above)
Comment (thestr4ng3r) [reply libr/core/rtr.c]: Not needed, the struct is not initialized in all libuv examples I have seen.

File: meson_options.txt [hunk 3]
```diff
@@ -20,3 +20,4 @@ option('use_sys_zlib', type: 'boolean', value: false)
 option('use_sys_lz4', type: 'boolean', value: false)
 option('use_sys_xxhash', type: 'boolean', value: false)
 option('use_sys_openssl', type: 'boolean', value: false)
+option('use_libuv', type: 'boolean', value: true)
```
Comment (pelijah) [inline meson_options.txt]: Maybe `use_sys_libuv`?

Comment (thestr4ng3r) [reply meson_options.txt]: There is no bundled alternative. If it's set to false, it doesn't use libuv at all.

**Comment by radare** [comment]
Please fix the conflict with configure. should be good to merge now i think

**Comment by thestr4ng3r** [comment]
rebased

**Pull request merged**